# Lending Club Loan Data 

## Dataset: rejected_2007_to_2018.csv

### Objetivo

Analizar la estructura, calidad y relaciones del dataset para comprender su papel dentro de la arquitectura Lakehouse y del modelo dimensional.

---

### Contexto del negocio

# Contexto del Negocio

El dataset Rejected Loans almacena información histórica de solicitudes de crédito que no fueron aprobadas por Lending Club entre 2007 y 2018.

A diferencia de Accepted Loans, este conjunto de datos representa decisiones negativas dentro del proceso de originación crediticia.

La información contiene características de los solicitantes, montos requeridos, ingresos declarados, nivel de endeudamiento y motivos asociados al rechazo de la solicitud.

Desde una perspectiva de negocio, este dataset permite analizar:

- Factores asociados al rechazo de créditos.
- Políticas de aprobación.
- Perfil de clientes de alto riesgo.
- Oportunidades de expansión de mercado.
- Sesgos potenciales en los procesos de evaluación.

El dataset resulta especialmente valioso para el desarrollo de modelos predictivos orientados a la evaluación de riesgo crediticio y optimización de estrategias de originación.

---

### Rol dentro de la arquitectura

# Contexto del Negocio

El dataset Rejected Loans almacena información histórica de solicitudes de crédito que no fueron aprobadas por Lending Club entre 2007 y 2018.

A diferencia de Accepted Loans, este conjunto de datos representa decisiones negativas dentro del proceso de originación crediticia.

La información contiene características de los solicitantes, montos requeridos, ingresos declarados, nivel de endeudamiento y motivos asociados al rechazo de la solicitud.

Desde una perspectiva de negocio, este dataset permite analizar:

- Factores asociados al rechazo de créditos.
- Políticas de aprobación.
- Perfil de clientes de alto riesgo.
- Oportunidades de expansión de mercado.
- Sesgos potenciales en los procesos de evaluación.

El dataset resulta especialmente valioso para el desarrollo de modelos predictivos orientados a la evaluación de riesgo crediticio y optimización de estrategias de originación.

## Relación con el Caso de Negocio

Accepted Loans representa el resultado positivo del proceso de evaluación crediticia.

Rejected Loans representa el resultado negativo del mismo proceso.

La combinación de ambos datasets permite analizar integralmente las decisiones de originación, facilitando el desarrollo de modelos predictivos de aprobación y riesgo crediticio.

# 1. Importación de librerías

Se cargan las librerías necesarias para el análisis exploratorio de datos.

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 200)

# 2. Carga del Dataset

Se carga el dataset para realizar el análisis exploratorio y comprender su estructura, calidad y relevancia dentro del dominio de negocio.

In [2]:
df = pd.read_csv("../../../data/raw/lendingClub/rejected_2007_to_2018Q4.csv")

print(f"Filas: {df.shape[0]}")
print(f"Columnas: {df.shape[1]}")

Filas: 27648741
Columnas: 9


# 3. Tamaño del Dataset

El dataset `Rejected Loans` contiene información histórica de solicitudes de crédito que no fueron aprobadas por Lending Club entre los años 2007 y 2018.

## Dimensiones Generales

| Métrica | Valor |
|----------|---------|
| Filas | 27,648,741 |
| Columnas | 9 |

---

## Interpretación

A diferencia del dataset Accepted Loans, este conjunto de datos presenta una estructura considerablemente más compacta, enfocada exclusivamente en la etapa de evaluación y rechazo de solicitudes crediticias.

Cada registro representa una solicitud de préstamo que no logró superar los criterios de aprobación establecidos por Lending Club.

---

## Características Generales

La información disponible se concentra principalmente en variables utilizadas durante el proceso de originación y evaluación de riesgo.

Esto permite analizar:

- Características de los solicitantes rechazados.
- Factores asociados a decisiones negativas.
- Patrones de riesgo.
- Comportamiento de las políticas de aprobación.
- Oportunidades potenciales de expansión crediticia.

---

## Diferencias Frente a Accepted Loans

Mientras Accepted Loans almacena información del ciclo completo del préstamo, Rejected Loans únicamente registra información disponible antes de la decisión crediticia.

| Accepted Loans | Rejected Loans |
|----------------|----------------|
| Créditos aprobados | Créditos rechazados |
| Pagos | No aplica |
| Mora | No aplica |
| Recuperación | No aplica |
| Desempeño financiero | No aplica |
| Evaluación inicial | Sí |

---

## Relevancia para el Negocio

Este dataset permite comprender qué perfiles de clientes fueron considerados demasiado riesgosos o no cumplieron los criterios de aprobación establecidos por la entidad.

Su análisis resulta especialmente útil para:

- Optimización de políticas de originación.
- Evaluación de estrategias de riesgo.
- Análisis de mercado potencial.
- Construcción de modelos de aprobación crediticia.

---

## Relevancia para Ingeniería de Datos

Aunque la estructura contiene únicamente nueve atributos, el volumen superior a 27 millones de registros convierte este dataset en una fuente masiva de información.

Por esta razón, futuras etapas de procesamiento deberán considerar:

- Procesamiento distribuido.
- Estrategias de particionamiento.
- Optimización de almacenamiento.
- Procesos eficientes de agregación.

---

## Relación con el Proyecto

Rejected Loans complementa la información contenida en Accepted Loans al incorporar el universo de solicitudes que no fueron aprobadas.

La combinación de ambos datasets permite analizar de forma integral el proceso de originación crediticia, proporcionando una visión completa de decisiones positivas y negativas dentro del negocio financiero.

# 4. Vista Preliminar

Se inspeccionan los primeros registros del dataset con el objetivo de comprender la naturaleza de la información almacenada y el nivel de granularidad de los datos.

## Muestra de Registros

| Amount Requested | Application Date | Loan Title | Risk Score | Debt-To-Income Ratio | Zip Code | State | Employment Length | Policy Code |
|-----------------|-----------------|------------|------------|---------------------|----------|-------|------------------|------------|
| 1000 | 2007-05-26 | Wedding Covered but No Honeymoon | 693 | 10% | 481xx | NM | 4 years | 0 |
| 1000 | 2007-05-26 | Consolidating Debt | 703 | 10% | 010xx | MA | < 1 year | 0 |
| 11000 | 2007-05-27 | Want to consolidate my debt | 715 | 10% | 212xx | MD | 1 year | 0 |
| 6000 | 2007-05-27 | waksman | 698 | 38.64% | 017xx | MA | < 1 year | 0 |
| 1500 | 2007-05-27 | mdrigo | 509 | 9.43% | 209xx | MD | < 1 year | 0 |

---

## Observaciones Iniciales

La estructura del dataset refleja información capturada durante la fase de evaluación crediticia previa a la aprobación o rechazo de una solicitud.

A diferencia de Accepted Loans, no se observan variables relacionadas con:

- Pagos.
- Mora.
- Recuperación de cartera.
- Estados posteriores del préstamo.

Esto confirma que la información fue registrada antes de que Lending Club tomara una decisión final sobre la solicitud.

---

## Variables Observadas

### Información de la Solicitud

- Amount Requested
- Application Date
- Loan Title

Estas variables describen la necesidad de financiamiento expresada por el solicitante.

---

### Información de Riesgo

- Risk Score
- Debt-To-Income Ratio

Representan indicadores utilizados durante la evaluación crediticia.

El Risk Score refleja una estimación de solvencia del solicitante, mientras que el Debt-To-Income Ratio (DTI) mide el nivel de endeudamiento respecto a los ingresos.

---

### Información Geográfica

- Zip Code
- State

Permiten identificar patrones regionales asociados a las solicitudes rechazadas.

---

### Información Laboral

- Employment Length

Proporciona una medida aproximada de estabilidad laboral del solicitante.

---

### Información Operativa

- Policy Code

Corresponde a una clasificación interna utilizada por Lending Club durante el proceso de evaluación.

---

## Hipótesis Inicial de Granularidad

A partir de los registros observados se plantea la siguiente hipótesis:

**1 registro = 1 solicitud rechazada**

Cada fila representa una solicitud individual que no logró cumplir los criterios de aprobación establecidos por la entidad.

Esta hipótesis será validada posteriormente durante el análisis de duplicados e identificadores.

---

## Diferencias Frente a Accepted Loans

El dataset Rejected Loans almacena exclusivamente información disponible antes de la decisión crediticia.

Por esta razón, el conjunto de atributos se encuentra orientado principalmente a:

- Perfil del solicitante.
- Riesgo estimado.
- Motivo de financiamiento.
- Variables utilizadas en la evaluación inicial.

---

## Relevancia para el Negocio

Este dataset permite estudiar los perfiles de clientes que no lograron acceder al crédito, proporcionando información valiosa para:

- Evaluar políticas de aprobación.
- Analizar oportunidades de mercado.
- Identificar segmentos subatendidos.
- Construir modelos de decisión crediticia.

---

## Conclusión Preliminar

Rejected Loans constituye una fuente especializada en el análisis de originación crediticia, permitiendo comprender las características de los solicitantes rechazados y los factores potencialmente asociados a decisiones negativas dentro del proceso de evaluación.

In [3]:
df.head()

,Amount Requested,Application Date,Loan Title,Risk_Score,Debt-To-Income Ratio,Zip Code,State,Employment Length,Policy Code
0,1000.0,2007-05-26,Wedding Covered but No Honeymoon,693.0,10%,481xx,NM,4 years,0.0
1,1000.0,2007-05-26,Consolidating Debt,703.0,10%,010xx,MA,< 1 year,0.0
2,11000.0,2007-05-27,Want to consolidate my debt,715.0,10%,212xx,MD,1 year,0.0
3,6000.0,2007-05-27,waksman,698.0,38.64%,017xx,MA,< 1 year,0.0
4,1500.0,2007-05-27,mdrigo,509.0,9.43%,209xx,MD,< 1 year,0.0


# 5. Estructura General

Se analiza la composición estructural del dataset con el fin de comprender la naturaleza de las variables disponibles y evaluar su utilidad para futuros procesos analíticos relacionados con riesgo crediticio y originación.

---

## Resumen Estructural

| Métrica | Valor |
|----------|---------|
| Registros | 27,648,741 |
| Columnas | 9 |
| Variables Numéricas | 3 |
| Variables Categóricas | 6 |
| Tipos de Datos Distintos | 2 |

---

## Distribución de Tipos de Datos

| Tipo de Dato | Cantidad |
|--------------|----------|
| string | 6 |
| float64 | 3 |

---

## Interpretación

A diferencia de Accepted Loans, este dataset presenta una estructura significativamente más compacta y especializada.

La información disponible se concentra exclusivamente en variables utilizadas durante la evaluación inicial de solicitudes de crédito.

No se observan atributos relacionados con:

- Pagos.
- Estados del préstamo.
- Mora.
- Recuperación de cartera.
- Seguimiento financiero.

Esto confirma que el dataset representa una etapa previa a la aprobación o desembolso del crédito.

---

## Componentes Principales

### Variables de Solicitud

Incluyen información asociada a la necesidad de financiamiento expresada por el cliente.

Ejemplos:

- Amount Requested
- Loan Title
- Application Date

Estas variables describen la naturaleza y características de la solicitud presentada.

---

### Variables de Riesgo

Incluyen atributos utilizados durante la evaluación crediticia.

Ejemplos:

- Risk_Score
- Debt-To-Income Ratio

Estas variables permiten estimar la capacidad de pago y el perfil de riesgo del solicitante.

---

### Variables Geográficas

Ejemplos:

- State
- Zip Code

Permiten realizar análisis regionales y detectar patrones geográficos asociados a solicitudes rechazadas.

---

### Variables Laborales

Ejemplos:

- Employment Length

Aportan información relacionada con la estabilidad laboral del solicitante.

---

### Variables Operativas

Ejemplos:

- Policy Code

Representan atributos internos utilizados por Lending Club durante el proceso de originación.

---

## Complejidad del Dataset

Aunque el volumen supera los 27 millones de registros, la estructura de nueve columnas facilita considerablemente las tareas de procesamiento y análisis.

Esta simplicidad permite:

- Procesos de limpieza más rápidos.
- Menor complejidad de transformación.
- Mayor facilidad para generar modelos predictivos.
- Integración sencilla con otras fuentes crediticias.

---

## Implicaciones para Ingeniería de Datos

El principal desafío de este dataset no radica en la cantidad de atributos, sino en el volumen masivo de registros.

Por esta razón, será necesario considerar:

- Procesamiento distribuido.
- Estrategias de particionamiento.
- Optimización de consultas.
- Almacenamiento eficiente.

---

## Hipótesis Inicial de Modelado

La estructura observada sugiere la construcción de entidades asociadas a:

### Dimensiones

- Tiempo
- Geografía
- Perfil Laboral
- Riesgo Crediticio

### Hechos

- Solicitudes Rechazadas

---

## Relación con el Negocio

El dataset captura la información mínima necesaria para comprender por qué determinados solicitantes no lograron acceder a un crédito.

Esta perspectiva complementa la información de Accepted Loans y permite analizar integralmente el proceso de originación crediticia.

---

## Conclusión

Rejected Loans presenta una estructura simple pero altamente especializada, enfocada exclusivamente en la etapa de evaluación de solicitudes.

Su combinación de gran volumen y baja dimensionalidad lo convierte en una fuente ideal para estudios de riesgo crediticio, segmentación de clientes y optimización de políticas de aprobación.

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 27648741 entries, 0 to 27648740
Data columns (total 9 columns):
 #   Column                Dtype  
---  ------                -----  
 0   Amount Requested      float64
 1   Application Date      str    
 2   Loan Title            str    
 3   Risk_Score            float64
 4   Debt-To-Income Ratio  str    
 5   Zip Code              str    
 6   State                 str    
 7   Employment Length     str    
 8   Policy Code           float64
dtypes: float64(3), str(6)
memory usage: 3.0 GB


In [5]:
df.dtypes.value_counts()

str        6
float64    3
Name: count, dtype: int64

# 6. Identificación de Claves

Se identifican:

- Clave primaria
- Claves foráneas
- Relaciones con otros datasets

In [ ]:
df.head(1)

In [6]:
df.duplicated().sum()

np.int64(157954)

## Análisis Inicial

A diferencia de Accepted Loans, el dataset Rejected Loans no incorpora un identificador único explícito para cada solicitud.

No se observa una columna equivalente a:

- id
- loan_id
- application_id

Por lo tanto, no existe una clave primaria natural que permita identificar de forma inequívoca cada registro.

---

## Implicaciones

La ausencia de un identificador único dificulta:

- La trazabilidad individual de solicitudes.
- La detección precisa de registros repetidos.
- La construcción de relaciones directas con otras fuentes de datos.

---

## Posibles Claves de Negocio

Se identifican varias columnas que podrían participar en la construcción de una clave compuesta lógica:

- Application Date
- Amount Requested
- Loan Title
- State
- Zip Code

Sin embargo, ninguna de ellas garantiza unicidad de forma individual.

---

## Granularidad del Dataset

A partir de la información observada se plantea la siguiente definición de grano:

**1 registro = 1 solicitud rechazada**

Cada fila representa una solicitud de crédito que no logró superar los criterios de aprobación definidos por Lending Club.

---

## Consideraciones para el Modelado

Durante procesos posteriores de transformación será recomendable generar una clave sustituta (Surrogate Key) que permita identificar de manera única cada registro procesado dentro de la arquitectura analítica.

---

## Conclusión

Rejected Loans no dispone de una clave primaria natural. Por esta razón, futuras etapas de integración deberán considerar mecanismos alternativos de identificación y trazabilidad de registros.

# 7. Calidad de Datos

Se analiza la completitud de la información y posibles problemas de calidad.

In [7]:
null_percentage = (
    df.isnull()
      .mean()
      .mul(100)
      .sort_values(ascending=False)
)

null_percentage.head(20)

Risk_Score              66.902251
Employment Length        3.440862
Loan Title               0.004720
Policy Code              0.003320
Zip Code                 0.001060
State                    0.000080
Amount Requested         0.000000
Debt-To-Income Ratio     0.000000
Application Date         0.000000
dtype: float64

In [8]:
high_nulls = null_percentage[null_percentage > 50]

medium_nulls = null_percentage[
    (null_percentage > 20)
    & (null_percentage <= 50)
]

low_nulls = null_percentage[
    (null_percentage > 0)
    & (null_percentage <= 20)
]

print("Columnas > 50%:", len(high_nulls))
print("Columnas entre 20%-50%:", len(medium_nulls))
print("Columnas < 20%:", len(low_nulls))

Columnas > 50%: 1
Columnas entre 20%-50%: 0
Columnas < 20%: 5


## Interpretación

Se realizó un análisis de completitud con el objetivo de identificar la presencia de valores faltantes y evaluar su impacto potencial sobre los procesos analíticos posteriores.

---

## Resumen General

| Categoría | Cantidad |
|------------|----------|
| Columnas > 50% nulos | 1 |
| Columnas entre 20% y 50% | 0 |
| Columnas < 20% | 5 |

---

## Variables con Valores Faltantes

| Variable | % Nulos |
|------------|---------|
| Risk_Score | 66.90% |
| Employment Length | 3.44% |
| Loan Title | 0.00% |
| Policy Code | 0.00% |
| Zip Code | 0.00% |
| State | 0.00% |

---

## Interpretación de Negocio

A diferencia de Accepted Loans, el dataset Rejected Loans presenta niveles muy bajos de valores faltantes.

La mayoría de los atributos esenciales para el análisis de solicitudes rechazadas se encuentran prácticamente completos.

---

### Risk_Score

La variable Risk_Score presenta aproximadamente un 66.9% de valores faltantes.

Este comportamiento sugiere que el puntaje de riesgo no estuvo disponible para una gran proporción de solicitudes o que su captura dependía de procesos específicos de evaluación.

Desde una perspectiva analítica, esta columna deberá evaluarse cuidadosamente antes de ser utilizada en modelos predictivos o segmentaciones.

---

### Employment Length

La variable Employment Length presenta únicamente un 3.44% de valores faltantes.

Este nivel de ausencia resulta bajo y no representa una amenaza significativa para los análisis posteriores.

---

### Variables Operativas

Las columnas:

- Amount Requested
- Application Date
- Debt-To-Income Ratio

presentan completitud total.

Esto resulta especialmente valioso debido a que representan atributos fundamentales dentro del proceso de evaluación crediticia.

---

## Hallazgos Relevantes

### Alta Completitud

El dataset mantiene una calidad estructural considerablemente superior a la observada en Accepted Loans.

La mayoría de los atributos presentan niveles de completitud cercanos al 100%.

---

### Riesgo como Variable Opcional

La ausencia masiva de valores en Risk_Score sugiere que este atributo no fue generado para todas las solicitudes.

Por lo tanto, su utilización deberá ser evaluada según el período histórico o la disponibilidad efectiva de información.

---

### Calidad Operativa

La evidencia observada indica que los procesos de captura de información para solicitudes rechazadas mantienen un alto nivel de consistencia y completitud.

---

## Implicaciones para la Arquitectura Lakehouse

### Bronze

Se conservarán todos los atributos originales para mantener la trazabilidad completa de la fuente.

### Silver

Será necesario evaluar estrategias de tratamiento para Risk_Score debido a su elevado porcentaje de valores faltantes.

### Gold

Las variables con alta completitud podrán utilizarse directamente para análisis de riesgo, segmentación y evaluación de políticas crediticias.

---

## Conclusión

Rejected Loans presenta una calidad de datos notablemente alta para una fuente de gran volumen.

La única excepción significativa corresponde a Risk_Score, cuya ausencia en más del 60% de los registros requerirá una evaluación específica durante las etapas de transformación y modelado analítico.

# 8. Detección de Registros Duplicados

In [ ]:
df.duplicated().sum()

# 8. Detección de Registros Duplicados

## Interpretación

Se realizó una validación de duplicidad sobre la totalidad de los registros presentes en el dataset.

---

## Resultados

| Métrica | Valor |
|----------|---------|
| Registros Totales | 27,648,741 |
| Registros Duplicados | 157,954 |
| Porcentaje Aproximado | 0.57% |

---

## Hallazgos

Se identificó la presencia de registros duplicados dentro del dataset.

Aunque el porcentaje relativo es bajo respecto al volumen total de información, la cantidad absoluta de registros duplicados resulta significativa debido al tamaño de la fuente.

---

## Posibles Causas

Los duplicados pueden estar asociados a:

- Solicitudes registradas múltiples veces.
- Procesos de consolidación histórica.
- Actualizaciones operativas.
- Replicación de información durante procesos de carga.

---

## Impacto Analítico

La existencia de registros duplicados puede generar:

- Sobreestimación de solicitudes rechazadas.
- Distorsión de métricas de riesgo.
- Sesgos en modelos predictivos.
- Incremento artificial de frecuencias por segmento.

---

## Recomendaciones

Durante la capa Silver se recomienda:

- Identificar reglas de deduplicación.
- Construir claves compuestas de negocio.
- Evaluar el impacto de los registros repetidos sobre las métricas principales.

---

## Conclusión

El dataset presenta un nivel reducido de duplicidad relativa (0.57%), aunque el volumen absoluto de registros repetidos justifica la implementación de controles de calidad y estrategias de deduplicación en futuras etapas del procesamiento.

# 9. Variables de Negocio Relevantes

Con el objetivo de comprender los factores asociados a las solicitudes rechazadas, se seleccionó un conjunto de variables consideradas estratégicas dentro del proceso de evaluación crediticia.

---

## Vista de Variables de Negocio

| Amount Requested | Risk_Score | Debt-To-Income Ratio | Employment Length | State |
|-----------------|------------|---------------------|------------------|-------|
| 1000.0 | 693.0 | 10% | 4 years | NM |
| 1000.0 | 703.0 | 10% | < 1 year | MA |
| 11000.0 | 715.0 | 10% | 1 year | MD |
| 6000.0 | 698.0 | 38.64% | < 1 year | MA |
| 1500.0 | 509.0 | 9.43% | < 1 year | MD |

---

## Interpretación de las Variables

### Amount Requested

Representa el monto solicitado por el cliente.

Permite analizar:

- Necesidades de financiamiento.
- Distribución de solicitudes.
- Relación entre monto solicitado y probabilidad de rechazo.

Pregunta de negocio:

¿Qué montos son más frecuentes entre las solicitudes rechazadas?

---

### Risk_Score

Corresponde a una medida de riesgo asociada al solicitante.

Esta variable constituye uno de los principales indicadores utilizados durante la evaluación crediticia.

Permite analizar:

- Distribución de perfiles de riesgo.
- Relación entre riesgo y rechazo.
- Segmentación de solicitantes.

Pregunta de negocio:

¿Los clientes rechazados presentan puntajes de riesgo inferiores a los clientes aprobados?

---

### Debt-To-Income Ratio

Representa la relación entre deuda e ingresos del solicitante.

Este indicador es ampliamente utilizado en procesos de evaluación crediticia debido a que refleja la capacidad financiera del cliente para asumir nuevas obligaciones.

Pregunta de negocio:

¿Existe un nivel de endeudamiento a partir del cual aumenta significativamente la probabilidad de rechazo?

---

### Employment Length

Describe la antigüedad laboral declarada por el solicitante.

La estabilidad laboral suele considerarse un indicador indirecto de estabilidad financiera.

Pregunta de negocio:

¿Los solicitantes con menor antigüedad laboral presentan mayores tasas de rechazo?

---

### State

Indica el estado de residencia del solicitante.

Esta variable permite realizar análisis geográficos y detectar diferencias regionales en las decisiones de aprobación.

Pregunta de negocio:

¿Existen estados con mayores niveles de rechazo que otros?

---

## Agrupación por Procesos de Negocio

### Solicitud de Crédito

Variables:

- Amount Requested
- Loan Title
- Application Date

Pregunta:

¿Qué tipo de financiamiento buscan los clientes?

---

### Evaluación de Riesgo

Variables:

- Risk_Score
- Debt-To-Income Ratio
- Employment Length

Pregunta:

¿Qué características financieras presentan los solicitantes rechazados?

---

### Segmentación Geográfica

Variables:

- State
- Zip Code

Pregunta:

¿Existen patrones regionales asociados al rechazo de solicitudes?

---

## Hallazgos Iniciales

A partir de los registros observados se identifican algunos patrones preliminares:

- Los montos solicitados presentan alta variabilidad.
- Los solicitantes muestran diferentes niveles de riesgo crediticio.
- La estabilidad laboral parece variar significativamente entre registros.
- Existen diferencias importantes en los niveles de endeudamiento reportados.

---

## Relevancia para el Proyecto

Estas variables permiten construir análisis relacionados con:

- Originación crediticia.
- Políticas de aprobación.
- Segmentación de clientes.
- Evaluación de riesgo.
- Modelos predictivos de aprobación y rechazo.

---

## Conclusión

Las variables analizadas representan los principales factores considerados durante la evaluación de solicitudes crediticias y constituyen la base para comprender los perfiles de clientes que no lograron acceder a financiamiento dentro de la plataforma Lending Club.

In [9]:
important_cols = [
    "Amount Requested",
    "Risk_Score",
    "Debt-To-Income Ratio",
    "Employment Length",
    "State"
]

df[important_cols].head()

,Amount Requested,Risk_Score,Debt-To-Income Ratio,Employment Length,State
0,1000.0,693.0,10%,4 years,NM
1,1000.0,703.0,10%,< 1 year,MA
2,11000.0,715.0,10%,1 year,MD
3,6000.0,698.0,38.64%,< 1 year,MA
4,1500.0,509.0,9.43%,< 1 year,MD


# 10. Variables Categóricas

In [10]:
categorical_columns = df.select_dtypes(include=["object"])

categorical_columns.columns.tolist()

C:\Users\ferna\AppData\Local\Temp\ipykernel_35348\1644596.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include=["object"])


['Application Date',
 'Loan Title',
 'Debt-To-Income Ratio',
 'Zip Code',
 'State',
 'Employment Length']

In [11]:
for col in categorical_columns.columns:
    print(f"{col}: {df[col].nunique()}")

Application Date: 4238
Loan Title: 73927
Debt-To-Income Ratio: 126145
Zip Code: 1001
State: 51
Employment Length: 11


Con el objetivo de comprender la diversidad de categorías presentes en el dataset, se analizó la cardinalidad de las variables categóricas disponibles.

---

## Cardinalidad de Variables Categóricas

| Variable | Valores Únicos |
|-----------|-----------|
| Application Date | 4,238 |
| Loan Title | 73,927 |
| Debt-To-Income Ratio | 126,145 |
| Zip Code | 1,001 |
| State | 51 |
| Employment Length | 11 |

---

## Interpretación

Las variables categóricas presentan distintos niveles de granularidad, permitiendo identificar atributos operativos, geográficos y financieros utilizados durante la evaluación de solicitudes.

---

### Baja Cardinalidad

Variables con pocas categorías y alto valor analítico.

#### Employment Length (11)

Representa la antigüedad laboral del solicitante.

Permite analizar:

- Estabilidad laboral.
- Relación entre experiencia laboral y rechazo.
- Segmentación de clientes.

#### State (51)

Representa los estados de residencia de los solicitantes.

Permite:

- Análisis geográfico.
- Comparaciones regionales.
- Identificación de patrones territoriales.

---

### Cardinalidad Media

#### Zip Code (1,001)

Proporciona un mayor nivel de detalle geográfico.

Puede utilizarse para:

- Segmentación regional.
- Estudios de concentración de solicitudes.
- Identificación de zonas de riesgo.

#### Application Date (4,238)

Permite realizar análisis temporales de largo plazo.

Facilita el estudio de:

- Tendencias históricas.
- Estacionalidad.
- Evolución de las políticas crediticias.

---

### Alta Cardinalidad

#### Loan Title (73,927)

Contiene descripciones libres proporcionadas por los solicitantes.

La elevada cardinalidad refleja una gran diversidad de necesidades de financiamiento.

Ejemplos:

- Consolidación de deuda.
- Bodas.
- Mejoras del hogar.
- Gastos personales.

#### Debt-To-Income Ratio (126,145)

Presenta una cardinalidad extremadamente elevada debido a la gran precisión de los valores reportados.

Aunque técnicamente se almacena como texto, representa una variable numérica continua relacionada con el nivel de endeudamiento del solicitante.

Se recomienda su conversión a formato numérico durante futuras etapas de transformación.

---

## Hallazgos Relevantes

### Variable de Mayor Valor Predictivo

La variable Debt-To-Income Ratio posee una granularidad muy elevada y representa uno de los principales indicadores utilizados en modelos de riesgo crediticio.

---

### Variables Geográficas

State y Zip Code permiten incorporar componentes espaciales al análisis de solicitudes rechazadas.

---

### Variables Textuales

Loan Title contiene información potencialmente valiosa para análisis de texto y categorización de motivos de solicitud.

---

## Conclusión

Rejected Loans presenta una estructura categórica relativamente sencilla, donde destacan las variables geográficas, laborales y financieras como principales fuentes de información para el análisis de solicitudes rechazadas.

# 11. Relaciones del Dataset

El análisis de las variables permite identificar relaciones relevantes dentro del proceso de evaluación crediticia previo a la decisión de aprobación o rechazo.

---

## Relación entre Monto Solicitado y Riesgo

Variables:

- Amount Requested
- Risk_Score

Interpretación:

Los montos solicitados pueden estar asociados a diferentes niveles de riesgo percibido por la entidad financiera.

Pregunta de negocio:

¿Los solicitantes con mayores montos presentan perfiles de riesgo más elevados?

---

## Relación entre Endeudamiento y Riesgo

Variables:

- Debt-To-Income Ratio
- Risk_Score

Interpretación:

El nivel de endeudamiento constituye uno de los principales factores considerados durante la evaluación crediticia.

Pregunta de negocio:

¿Los clientes con mayores niveles de endeudamiento presentan peores calificaciones de riesgo?

---

## Relación entre Estabilidad Laboral y Riesgo

Variables:

- Employment Length
- Risk_Score

Interpretación:

La estabilidad laboral suele utilizarse como indicador indirecto de estabilidad financiera.

Pregunta de negocio:

¿Los solicitantes con menor experiencia laboral presentan mayores probabilidades de rechazo?

---

## Relación entre Geografía y Solicitudes

Variables:

- State
- Zip Code
- Amount Requested

Interpretación:

La ubicación geográfica puede influir sobre las características de las solicitudes y los perfiles de riesgo observados.

Pregunta de negocio:

¿Existen diferencias regionales en los montos solicitados y niveles de riesgo?

---

## Relación Temporal

Variables:

- Application Date
- Amount Requested
- Risk_Score

Interpretación:

Permiten analizar la evolución histórica de las solicitudes rechazadas a lo largo del tiempo.

Pregunta de negocio:

¿Cómo evolucionó el perfil de riesgo de los solicitantes entre 2007 y 2018?

---

## Conclusión

Las relaciones identificadas se concentran principalmente en la fase de originación crediticia, permitiendo estudiar los factores asociados al rechazo de solicitudes y la evolución histórica de las políticas de evaluación.

# 12. Conclusiones Técnicas

El análisis exploratorio realizado sobre el dataset Rejected Loans permitió comprender la estructura, calidad y potencial analítico de la información histórica correspondiente a solicitudes de crédito rechazadas por Lending Club entre 2007 y 2018.

---

## Hallazgos Generales

El dataset contiene:

- 27,648,741 registros.
- 9 variables.
- Más de una década de información histórica.

La granularidad identificada corresponde a:

**1 registro = 1 solicitud rechazada**

---

## Calidad de Datos

Se identificó una calidad general elevada.

Resultados principales:

- 1 columna con más del 50% de valores faltantes.
- 0 columnas entre 20% y 50%.
- 5 columnas con menos del 20%.

La única variable con afectación significativa corresponde a:

- Risk_Score (66.9%)

---

## Duplicados

Se identificaron:

- 157,954 registros duplicados.

Lo anterior representa aproximadamente:

- 0.57% del total del dataset.

Aunque el porcentaje es reducido, se recomienda implementar procesos de deduplicación durante las etapas de transformación.

---

## Variables de Mayor Valor Analítico

### Solicitud

- Amount Requested
- Loan Title
- Application Date

### Riesgo

- Risk_Score
- Debt-To-Income Ratio

### Perfil Laboral

- Employment Length

### Geografía

- State
- Zip Code

---

## Potencial Analítico

El dataset permite desarrollar iniciativas relacionadas con:

- Modelos de aprobación crediticia.
- Predicción de rechazo.
- Segmentación de solicitantes.
- Optimización de políticas de originación.
- Análisis de riesgo.

---

## Relación con Accepted Loans

Rejected Loans complementa directamente la información contenida en Accepted Loans.

Mientras Accepted Loans describe los créditos otorgados, Rejected Loans permite comprender las características de los solicitantes que no lograron acceder a financiamiento.

La combinación de ambos conjuntos proporciona una visión integral del proceso de originación crediticia.

---

## Implicaciones para Ingeniería de Datos

Debido al gran volumen de registros, futuras etapas requerirán:

- Procesamiento distribuido.
- Estrategias de particionamiento.
- Control de duplicados.
- Estandarización de variables.

Asimismo, será necesario transformar Debt-To-Income Ratio a formato numérico para facilitar análisis posteriores.

---

## Relación con la Arquitectura Lakehouse

Dentro de la arquitectura propuesta, Rejected Loans funcionará como una fuente complementaria para el análisis de originación crediticia.

Su flujo esperado será:

Rejected Loans

↓

Bronze

↓

Silver

↓

Intermediate

↓

Gold

↓

Diamond

---

## Conclusión Final

Rejected Loans constituye una fuente altamente especializada para el análisis de decisiones crediticias negativas.

Su combinación de gran volumen, estructura simple y alta calidad de datos lo convierte en un activo valioso para comprender los factores asociados al rechazo de solicitudes y complementar los análisis desarrollados sobre créditos aprobados.